---
title: ""
output: html_document
params:
  sample: ""
  list_genes: ""
  cutoff_sum_reads: ""
  FC: ""
  num_to_plot: ""
  metadata: ""
---

<style type="text/css">
  body{
  font-size: 15pt;
}
</style>

```{r echo=FALSE, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(warning  = FALSE)

```


```{r echo=FALSE, include=FALSE }

# load libraries

library(dplyr)
library(tidyr)
library(readr)
library(purrr)
library(tibble)
library(stringr)
#library("tidyverse")
library(data.table)
library(DESeq2)
library("EnhancedVolcano")
library(foreach)
#library("org.Hs.eg.db")


#orgdb = "org.Hs.eg.db"

# load functions
source("/nadata/users/racheli/nextflow/scripts/preprocess_data.R")

```


```{r}
sample = params$sample
list_genes = params$list_genes
cutoff_sum_reads = params$cutoff_sum_reads
FC = params$FC
num_to_plot = params$num_to_plot
metadata = params$metadata
```


```{r}
# params for DESeq analysis filtering
# sample = "Shefer"
# list_genes = ""
# 
# indir = paste0("/PostExome/RNAseq/Analysis/",sample,"_merged")
# out_dir = paste0(indir,"/deseq_res")
# dir.create(out_dir)
# out_file = paste0(out_dir,"/",sample,"_DE_report.html")
# cutoff_sum_reads = 50
# FC = 2
# num_to_plot = 10
# metadata=paste0("/PostExome/RNAseq/Analysis/",sample,"/list_samples_labeled.txt")
```

```{r}
# run female vs male - drop gene with associated to the sex
# sample = "male_vs_female"
# list_genes = ""
# # 
# indir = "/PostExome/RNAseq/Analysis"
# out_dir = paste0(indir,"/deseq_m_vs_f")
# dir.create(out_dir)
# out_file = paste0(out_dir,"/",sample,"_DE_report.html")
# cutoff_sum_reads = 50
# FC = 2
# num_to_plot = 10
# metadata="/PostExome/RNAseq/Analysis/male_vs_female_healthy.tsv"
```


## Report for project: **`r sample`**

```{r  echo=FALSE, include=FALSE}
# Get metadata:
sample_info <- readr::read_delim(file = metadata, delim = "\t", col_names = c("sample","condition"), show_col_types = FALSE)

sample_info$condition <- as.factor(sample_info$condition)
```

```{r}
# sample <- sample_info$sample
# rnaseqc_files <-c()
# for(s in sample){
#   family <- gsub("\\-.*","",s)
#   path <- paste0(indir,"/",family,"_merged/rnaseQC/",s,"_md.bam.gene_reads.gct")
#   rnaseqc_files <- c(rnaseqc_files,path)
#  
# }

# process_gct_file <- function(gene_count_file, sample_name) {
#   gene_count_file %>% read.delim(skip = 2) %>% group_by(Description) %>% summarise(Counts=sum(Counts)) %>% column_to_rownames(var = "Description") %>% setNames(c(sample_name))
# }
# 
# count_data <- purrr::map2_dfc(rnaseqc_files, sample, ~ process_gct_file(.x, .y))

```


```{r}
# Get counts and create dds
# Make sure that first column has the sample names, that these samples are all in countData, and that they are in the same order

rnaseqc_directory_family <- paste0(indir,"/rnaseQC")
gene_count_files <- list.files(rnaseqc_directory_family, pattern = "\\.gene_reads.gct$", full.names = TRUE)
gene_count_sample <- list.files(rnaseqc_directory_family, pattern = "\\.gene_reads.gct$", full.names = FALSE)
gene_count_sample <- sub("_md\\.bam\\.gene_reads\\.gct$", "", gene_count_sample)
```


```{r}
process_gct_file <- function(gene_count_file, sample_name) {
  gene_count_file %>% read.delim(skip = 2) %>% group_by(Description) %>% summarise(Counts=sum(Counts)) %>% column_to_rownames(var = "Description") %>% setNames(c(sample_name))

}

count_data <- purrr::map2_dfc(gene_count_files, gene_count_sample, ~ process_gct_file(.x, .y))


```


```{r}
ord <- order(match(sample_info$sample, names(count_data)))
sample_info <- sample_info[ord,]

colData = data.frame(row.names = names(count_data),
                     condition = sample_info$condition)
```

Samples list:

```{r}
colData
```


```{r}

# Create DESeq object and filter low expresses genes

dds <- DESeqDataSetFromMatrix(countData = count_data, colData = colData, design = ~ condition)

dds <- dds[ rowSums(counts(dds)) > cutoff_sum_reads , ]

```

## PCA


```{r echo = F}
# Plot PCA
if(nrow(colData) < 30){
	transformation <- rlog(dds, blind=FALSE)
}else{
	transformation <- vst(dds, blind=FALSE)
}
p = plotPCA(transformation, intgroup=c( "condition"))
plot(p + theme_classic(base_size = 16) + geom_point(size = 4) + geom_text(aes(label=p$data$name), size = 3, vjust=-1.3)+ theme(axis.text=element_text(colour="black", size = 18)))+ ylim(p$data$PC2[1],p$data$PC2[3]+1)

```



```{r}

# Function for specific gene expression

plot_gene = function(dds,gene_name,intgroup, info = ""){
  gene = plotCounts(dds, gene_name ,intgroup, returnData = T)
  p = ggplot(gene, aes(x=condition, y = count, color = condition )) + geom_point(size = 5, alpha=0.7) +
    ylim(0,max(gene$count)) + theme_bw(base_size = 15) + xlab("") +
    ggtitle(paste0(gene_name,info)) 
  plot(p)
}

```


## Expression for selected genes

```{r}
intgroup=c("condition")
if (!list_genes == "" ){
  for (gene in list_genes){
    plot_gene(dds,gene,intgroup)
  }
}

```


## DE analysis


```{r}

# DESeq analysis and write to file sig genes analysis

dds = DESeq(dds)

resultsNames(dds)

a = sig_genes(dds,FC = FC,pval=0.05)
sig = a[[1]]
res = a[[2]]


write.csv(res, paste0(out_dir,"/All_res_family.csv"))
write.csv(sig, paste0(out_dir,"/sig_genes_family.csv"))

```


parameters: 

FC >= `r FC`

pval (fdr corrected) = 0.05

**We get `r nrow(sig)` genes significantly changing**


Volcano plot for padj < 0.05 &  log2(FC) > `r log2(FC)`


```{r}

p=EnhancedVolcano(res,
                  lab = rownames(res),
                  x = 'log2FoldChange',
                  y = 'padj',
                  title = "",
                  pCutoff = 0.05,
                  FCcutoff = log2(FC),
                  pointSize = 0.5,
                  labSize = 2.5,
                  col = c("indianred1", "gold2", "cornflowerblue", "gray47"))
p3= p + theme_bw()  
plot(p3)
```



```{r include= FALSE}
# # GO


##Go term enrichment analysis for up/down regulated


#run_go1 = function(list_genes,universe, padj = "fdr", ont ="ALL", 
#                  simp = FALSE, info ="",orgdb= "org.Hs.eg.db", return_plot= F,
 #                 num_cat = 20){
  
  #ego2 <- enrichGO(gene         = list_genes,
                 #  OrgDb         =  orgdb,
                #   universe =  universe,
                 #  keyType       = 'SYMBOL',
                  # ont           = ont,
                  # pAdjustMethod = padj)
  #if (nrow(as.data.frame(ego2)) == 0) {
   # print(" no res")
    #return()
  #}
  
  #if (simp){
   # ego2 = simplify(ego2)
    
    
  #}
  
  #res_go = as.data.frame(ego2)
  #p = dotplot(ego2, showCategory=10)
  #p2 = (p + theme_bw(base_size = 10) + ggtitle(info))
  #plot(p2)
  
  #if (return_plot){
   # return(list(res_go,p2))
  #}
  #return(res_go)
#}

#res_go = run_go1(row.names(sig), row.names(res), padj = "fdr", ont ="ALL", orgdb= orgdb)
#write.csv(res_go, paste0(out_dir,"/GO_res.csv"))
```


## Expression for most significant genes

```{r}

sig = sig[order(sig$padj),]

if (nrow(sig) < 10 ){
  num_to_plot = nrow(sig)
}

if (nrow(sig) > 0){
  list_genes2 = row.names(sig)[1:num_to_plot]
  intgroup=c("condition")
  for (gene in list_genes2){
    plot_gene(dds,gene,intgroup)
  }
}

```

```{r}
library(pheatmap)

ntd <- normTransform(dds)
select <- order(rowMeans(counts(dds, normalized = TRUE)), decreasing = TRUE)[1:20]
pheatmap(assay(ntd)[select,], cluster_rows =TRUE, show_rownames = TRUE, cluster_cols = TRUE, annotation_col = colData)
```

